src/main/java/io/github/moulberry/notenoughupdates/profileviewer/GuiProfileViewer.java

```diff
@@ -1688,150 +1963,35 @@ private void renderXpBar(
-
-		if (Constants.PETS.has("custom_pet_leveling") &&
-			Constants.PETS.getAsJsonObject("custom_pet_leveling").has(pet_name)) {
-			JsonObject pet = Constants.PETS.getAsJsonObject("custom_pet_leveling").getAsJsonObject(pet_name);
-			if (pet.has("type") && pet.has("pet_levels")) {
-				int type = pet.get("type").getAsInt();
-				switch (type) {
-					case 1:
-						JsonArray defaultLevels = Constants.PETS.getAsJsonArray("pet_levels");
-						defaultLevels.addAll(pet.getAsJsonArray("pet_levels"));
-						petInfo.add("pet_levels", Constants.PETS.getAsJsonArray("pet_levels"));
-						break;
-					case 2:
-						petInfo.add("pet_levels", pet.getAsJsonArray("pet_levels"));
-						break;
-					default:
-						petInfo.add("pet_levels", Constants.PETS.getAsJsonArray("pet_levels"));
-						break;
-				}
-			} else {
-				petInfo.add("pet_levels", Constants.PETS.getAsJsonArray("pet_levels"));
-			}
-			if (pet.has("max_level")) {
-				petInfo.add("max_level", pet.get("max_level"));
-			} else {
-				petInfo.add("max_level", new JsonPrimitive(100));
-			}
-
-			if (pet.has("pet_rarity_offset")) {
-				petInfo.add("offset", pet.get("pet_rarity_offset"));
-			} else {
-				petInfo.add("offset", Constants.PETS.getAsJsonObject("pet_rarity_offset").get(rarity));
-			}
-
-		} else {
-			//System.out.println("Default Path");
-			petInfo.add("offset", Constants.PETS.getAsJsonObject("pet_rarity_offset").get(rarity));
-			petInfo.add("max_level", new JsonPrimitive(100));
-			petInfo.add("pet_levels", Constants.PETS.getAsJsonArray("pet_levels"));
-		}
-
-		return petInfo;
-
-	}
-
-	public static PetLevel getPetLevel(String pet_name, String rarity, float exp) {
-		JsonObject petInfo = getPetInfo(pet_name, rarity);
-		int offset = petInfo.get("offset").getAsInt();
-		int maxPetLevel = petInfo.get("max_level").getAsInt();
-		JsonArray levels = petInfo.getAsJsonArray("pet_levels");
-
-		float xpTotal = 0;
-		float level = 1;
-		float currentLevelRequirement = 0;
-		float currentLevelProgress = 0;
-
-		boolean addLevel = true;
-
-		for (int i = offset; i < offset + maxPetLevel - 1; i++) {
-			if (addLevel) {
-				currentLevelRequirement = levels.get(i).getAsFloat();
-				xpTotal += currentLevelRequirement;
-				if (xpTotal > exp) {
-					currentLevelProgress = (exp - (xpTotal - currentLevelRequirement));
-					addLevel = false;
-				} else {
-					level += 1;
-				}
-			} else {
-
-				xpTotal += levels.get(i).getAsFloat();
-
-			}
-		}
-
-		level += currentLevelProgress / currentLevelRequirement;
-		if (level <= 0) {
-			level = 1;
-		} else if (level > maxPetLevel) {
-			level = maxPetLevel;
-		}
-		PetLevel levelObj = new PetLevel();
-		levelObj.level = level;
-		levelObj.currentLevelRequirement = currentLevelRequirement;
-		levelObj.maxXP = xpTotal;
-		levelObj.levelPercentage = currentLevelProgress / currentLevelRequirement;
-		levelObj.levelXp = currentLevelProgress;
-		levelObj.totalXp = exp;
-		return levelObj;
+	private ItemStack getQuestionmarkSkull() {
```

Comment:
Im pretty sure this is just rearranging code, but we also have `Utils#createSkull` for this now.